### ISSUE TITLE
Shortcuts with newlines

### ISSUE THREAD

**Issue by psads-git** [mike-fabian/ibus-typing-booster#158]
Dear All,

Is it possible to expand a shortcut with a text that contains newlines?

Example: 

br

expanding to

"Best regards,

Paul"

Thanks in advance,

Paul

**Comment by mike-fabian** [comment]
I'll try to make that possible soon, it's on my TODO list.

**Comment by psads-git** [comment]
Thanks, Mike! That would be wonderful!

**Comment by mike-fabian** [comment]
Ah, one thing I am still thinking about in how to implement this is how to display multiline expansions in the candidate list.

Let's say you type “br” and this expands to 

> "Best regards,
>
> Paul"

Then this expansion can occupy only one slot in the candidate list, so it could look like this:


> 1 | Best regards\n\nPaul
> 2 | brain
> 3 | bridge
> ...

or like this:

> 1 | Best regards
>    | 
>    | Paul
> 2 | brain
> 3 | bridge
> ...

Which of these looks better?

Problems with that:

If the expansion is very long, this will look terrible, no matter whether it is displayed in  one line or multiple lines in the candidate list. So maybe for very long expansions, an abbreviation is needed. 

How to best show the abbreviation? 

> 1 | Start of expansion ... end of expansion

I.e. show the first 20 characters and the last 20 characters and insert ... in between

or 

> 1 | Start of expansion more stuff  ... 

I.e. Show the first 40 characters of the expansion and they cut and append a ...

or

> 1 | First line of expansion
>    | ... 
>    | last line of expansioin

and all the lines between the first and the last line are collapsed to ...

What do you think would look best?

I must experiment with that and see how that feels when using it.

I tend to think that somehow abbreviating it into one line would look best, multi-line candidates look terrible in some desktop environments. 

Maybe something like this would look OK:

> 1 | Start of first line ... [47 lines] ... end of last line

I.e. abbreviating the first and the last line to some resonable length and showing a count of the lines not shown in between.

Do you have any better ideas?

**Comment by mike-fabian** [comment]
This is how a multiline candidate  in the lookup table would look on Gnome (no matter whether Gnome Wayland or Gnome Xorg):

![Screenshot](https://user-images.githubusercontent.com/2330175/110370161-a1c6e080-804b-11eb-818d-597c5bcea759.png)

And this is how it would look like on XFCE (and most other non-Gnome desktops):

![Screenshot-xfce](https://user-images.githubusercontent.com/2330175/110370217-b3a88380-804b-11eb-97cc-7dccb6c6989a.png)

**Comment by mike-fabian** [comment]
The display on Gnome might be sort of acceptable, but even that would look bad if the expansion is much longer. 

On non-Gnome desktops like XFCE, multiline candidates are really terrible. There, the vertically tallest candidate forces all other candidates to be equally tall.

**Comment by mike-fabian** [comment]
This is how a multi line candidate would look like on Gnome when the option for a horizontal candidate list is used:

![Uploading Screenshot.png…]()

**Comment by psads-git** [comment]
Well, Mike, any user that uses shortcuts does not need to preview that in the suggestions lists. Thus, 

Shortcut br 

in the suggestions list would suffice. Do not you agree?

**Comment by psads-git** [comment]
Moreover, Mike, the user may want to have a shortcut for several paragraphs.

**Comment by mike-fabian** [comment]
But then one would need to remember the  expansion of the shortcuts. For example, if you have a shortcut "brp" expanding to 

Best regards,

Paul

and another one "brm" expanding to 

Best regards,

Mike

then typing the first 2 letters gives you both shortcuts. If you see only "brp" and "brm" I find it hard to remember after a while which was which. But maybe I could get used  to this. 

Or maybe a combination of both, i.e. display something like:

> brp ➡️ Best reg ... Paul
> brm ➡️ Best reg ... Mike

I.e. if one sees some text followed by ➡️ then the text before the ➡️ is the shortcut and the text behind the ➡️ gives some hint  what it expands to. 

How about that? Using an emoji like ➡️ saves some space compared to using a word like "Shortcut" and also makes it more stand out. And if the shortcut "brp" and "brm" is shown before the ➡️ then maybe showing a radical abbreviation of the expansion is enough as a reminder what the expansion is. 

Worring about how to display such multiline expansions of shortcuts so that it looks good and understandable was the main reason why I didn’t implement this yet.

It could be useful if one needs the same blocks of text very often, it might be useful to have such shortcuts expand to very long texts, maybe whole paragraphs. Then displaying all of it becomes impossible. But maybe the shorcut, a ➡️, and then an extremely abbreviated preview of the expansion might be good?

**Comment by mike-fabian** [comment]
Currently I use some shortcuts which are only a single line but about 80 characters long. It already looks quite weird when the candidate list becomes so wide that it is cut off by the size of the screen.

**Comment by psads-git** [comment]
I like your emoji solution, Mike!

Where I use shortcuts is for writing latex: Sometimes, the shortcut involves a lot of latex commands.

Your ibus-typing-booster already provides multilined shortcuts, if one paste those from gedit. However, for some strange reason, it does not work with Thunderbird.

**Comment by mike-fabian** [comment]
Yes, pasting from gedit works. The only work for me to do is:

1) Make it look better in the setup tool, i.e. make the expansion a Gtk.TextBuffer() (multiple lines) instead of a Gtk.Entry() (single line) and make it scrollable

2) Make it display as nice a possible in the candidate list.

Maybe I can first do 1) and do 2) later.  

Further problems are that if one has a lot of multiline shortcuts

- searching already defined shortcuts, maybe because one wants to  delete some of them again becomes  difficult, scrolling in a long list of multiline shortcuts is very tedious, maybe a search function is needed then.
- If one has huge amounts of shortcuts starting with many different letters, this will get in the way of completion of regular words. To avoid this, I start my shortcuts with an "ms" prefix because "ms" doesn't expand to anything in the languages I use. So I would use 'msbrp" for "Best regards, \n\n Paul". Maybe there is a better solution for that. Maybe allowing also some special letters at the beginning of such shortcuts. I am not sure. Currently, using something like \brp as a shortcut won’t work because most non-letter characters are cut away before searching for a completion. This is to make typing something like “(te” give expansions like “(testing” etc. when looking up in the dictionaries. Maybe using a short prefix like I do with “ms” common to all my shortcuts is good enough already? Or is there something I could do to improve that?
   
As you have found out by pasting from gedit, the basic functionality of multi-line shortcuts is already there, the work which needs to be done is improve the usability of this.

**Comment by mike-fabian** [comment]
> Your ibus-typing-booster already provides multilined shortcuts, if one pastes those from gedit. However, for some strange reason, it does not work with Thunderbird: the blank line between is cut.

That is really weird. I have never used Thunderbird so far. Only Thunderbird or did you find other such cases?

**Comment by psads-git** [comment]
In GMail in Firefox, it is still worse: No newline is inserted, Mike. However, in Chrome, no problem at all!

**Comment by mike-fabian** [comment]
> In GMail in Firefox, it is still worse: No newline is inserted, Mike. However, in Chrome, no problem at all!

I can reproduce that. When committing a multi-line text in GMail in Firefox, all new lines are stripped.
When doing the same in GMail in Chrome, it works, the newlines are correctly inserted.

Quite weird, It doesn't happen in all input fields in Firefox. For example here when writing a github comment I can 
commit a multi-line text like 

Best regards,

Mike 

without problems. 

I have no idea what could cause this, but it is certainly not the fault of ibus-typing-booster, ibus-typing-booster doesn't know whether one types in Chrome, Firefox, Github,... 

In all the following 3 cases: 

- Writing a new mail with GMail in Chrome
-  Writing a new mail with GMail in Firefox
- Writing a github comment in Firefox

I see in the debug log of ibus-typing-booster

```
2021-03-09 18:51:45,923 hunspell_table.py line 5395 do_set_content_type DEBUG: purpose=0 hints=0000000000000000
2021-03-09 18:51:45,923 hunspell_table.py line 5402 do_set_content_type DEBUG: self._input_purpose = 0 (InputPurpose.FREE_FORM)
```
 
So no input hints are set nd the input purpose is FREE_FORM in all 3 cases. 

As it makes no difference for ibus-typing-booster, the widget to edit a new mail in GMail in Firefox must be doing something special with multi-line commits from input methods. 

Nothing I can do to fix that, I think. 

Bug in GMail or in Firefox I guess.

**Comment by mike-fabian** [comment]
So I think I cannot do anything about the problem in Thunderbird either, ibus-typing-booster just calls

```
super().commit_text(IBus.Text.new_from_string(commit_phrase))
```

Where `commit_phrase` is a string which might contain anything, also newlines. 

What happens with that string then is out of control of ibus-typing-booster. It calls a commit function in ibus and ibus then commits that string. But I don't think ibus can do anything about that either. It is mysterious ☹.

**Comment by psads-git** [comment]
Thanks, Mike. I would suggest to commit line by line and not all lines at once. It may help.

**Comment by mike-fabian** [comment]
I'll experiment with that but that may be problematic as well. Many small commits in a short time seem to have the effect that some commits get lost. I am investigating what goes wrong with ibus-m17n when using Gnome Wayland for example in this bug: 

https://bugzilla.redhat.com/show_bug.cgi?id=1850832

When using these gu-itrans and hi-itrans methods with ibus-typing-booster, they work fine. Ibus-typing-booster waits until a word is finished and then commits the whole word. ibus-m17n commits as soon as one "unit" of characters is finished, this means it commits every 1-3 characters. And the some commits get lost ... That happens only in Wayland, it works in Xorg.

I am investigating at the moment whether I can work around that problem in ibus-m17n somehow, without success so far ...

When committing line by line there are two possibilities how to do it, 

1) commit each line separately and between each line commit a string which only has a newline and nothing else

2) commit each line separately and between each line use forward_key_event() to forward a Return key to the application.

2) tends to work better in my experience, but forward_key_event() is often problematic in Gnome Wayland as well.
In some versions of Qt it doesn't work at all (That's why ibus-typing-booster has this weird option "Use workaround for a but in Qt im module", enabling this option tries to avoid using forward_key_event() as much as possible.


Right now, in F34 Gnome Wayland, forward_key_event() seems broken again: 

https://bugzilla.redhat.com/show_bug.cgi?id=1936774

**Comment by mike-fabian** [comment]
I just tried this:

```
diff --git a/engine/hunspell_table.py b/engine/hunspell_table.py
index de13d95..0a22435 100644
--- a/engine/hunspell_table.py
+++ b/engine/hunspell_table.py
@@ -2493,8 +2493,9 @@ class TypingBoosterEngine(IBus.Engine):
                             'surrounding_text = '
                             '[text = "%s", cursor_pos = %s, anchor_pos = %s]',
                             text, cursor_pos, anchor_pos)
-        super().commit_text(
-            IBus.Text.new_from_string(commit_phrase))
+        for commit_line in commit_phrase.splitlines(keepends=True):
+            super().commit_text(
+                IBus.Text.new_from_string(commit_line))
         self._clear_input_and_update_ui()
         self._commit_happened_after_focus_in = True
         if self._off_the_record or self._hide_input:
```

and it does not help in Gmail in firefox.

When the commit contains `'Best regards,\n\n\n\nMike'` then this commits the following 4 chunks:

```
>>> 'Best regards,\n\n\n\nMike'.splitlines(keepends=True)
['Best regards,\n', '\n', '\n', '\n', 'Mike']
```

So one should see 3 empty lines between “Best regards,” and “Mike”. In gedit this works but in gmail in firefox one sees 4 spaces between “Best regards,” and “Mike”. Looks like this gmail in firefox  thing converts every newline in a commit string into a space, no matter whether the whole thing  is committed at once or in chunks. 

That leaves us with forward_key_event() ...

**Comment by mike-fabian** [comment]
Now I tried the forward_key_event() idea like this:

```
$ git diff 
diff --git a/engine/hunspell_table.py b/engine/hunspell_table.py
index de13d95..a8d7a7a 100644
--- a/engine/hunspell_table.py
+++ b/engine/hunspell_table.py
@@ -2493,8 +2493,25 @@ class TypingBoosterEngine(IBus.Engine):
                             'surrounding_text = '
                             '[text = "%s", cursor_pos = %s, anchor_pos = %s]',
                             text, cursor_pos, anchor_pos)
-        super().commit_text(
-            IBus.Text.new_from_string(commit_phrase))
+        if self._qt_im_module_workaround:
+            super().commit_text(
+                IBus.Text.new_from_string(commit_phrase))
+        else:
+            for commit_line in commit_phrase.splitlines(keepends=True):
+                if not commit_line.endswith('\n'):
+                    super().commit_text(
+                        IBus.Text.new_from_string(commit_line))
+                    continue
+                super().commit_text(
+                    IBus.Text.new_from_string(commit_line[:-1]))
+                self.forward_key_event(
+                    IBus.KEY_Return,
+                    self._keyvals_to_keycodes.ibus_keycode(IBus.KEY_Return),
+                    0)
+                # The sleep is needed because this is racy, without the
+                # sleep it is likely that all the commits come first
+                # followed by all the forwarded Return keys:
+                time.sleep(self._ibus_event_sleep_seconds)
         self._clear_input_and_update_ui()
         self._commit_happened_after_focus_in = True
         if self._off_the_record or self._hide_input:
```

This works in gedit *and* in GMail in firefox. On my it  desktop on Fedora 33.

Note that the sleep is necessary,  when committing something like `'Best regards,\n\n\n\nMike'` without the sleep the  result in gedit and Gmail in firefox is `'Best regards,Mike\n\n\n\n'`, i.e. all commits come first and all forwarded Returns come after all committed strings. 

With
 
```
        self._ibus_event_sleep_seconds = 0.1
```

The order was always right when I tested. But that might not work anymore when the system load is high, so I don't really like that hack.

But I see no alternative at the moment. At least with that hack it works in more cases than without and it does not much harm.

- When strings without newlines are committed, nothing is changed (i.e. Users who don't use multi-line shortcuts should see no difference at all in behaviour, no matter whether forward_key_event() works or not)
- When multiline shortcuts are used and forward_key_event() works, this makes ist work in GMail in Firefox (and probably in Thunderbird as well) and it still works in gedit (although you can see the slight slowdown introduced by the 0.1s sleeps between the lines).
- When multiline shortcuts are used and forward_key_event() is broken, one has to use the option “Use a workaround for a bug in Qt im module” (which I should really rename to “Avoid forwarding key events” ...) and when using that option it works as before, i.e. works in gedit but not in gmail in firefox.
-

**Comment by psads-git** [comment]
Thanks, Mike. I do not know what is the mechanism that AutoKey,

[AutoKey](https://github.com/autokey/autokey) ,

uses. However, multilined AutoKey's shortcuts expand correctly both in Thunderbird and GMail in Firefox. Maybe you could have a look at AutoKey for getting further ideas.

**Comment by mike-fabian** [comment]
> Thanks, Mike. I do not know what is the mechanism that AutoKey,
> 
> [AutoKey](https://github.com/autokey/autokey) ,
> 
> uses. However, multilined AutoKey's shortcuts expand correctly both in Thunderbird and GMail in Firefox. Maybe you could have a look at AutoKey for getting further ideas.


Auto key does not seem to be an ibus input method, so probably it cannot help me with problems  committing strings and forwarding key events via the ibus API.

**Comment by mike-fabian** [comment]
I wrote:

> brp ➡️ Best reg … Paul 

and you liked that idea:

>  I like your emoji solution, Mike!

But it looks like I cannot do that so easy without extra database lookups. Maybe later. 

Until now I made only a few small changes to make it more usable with multiline shortcuts:

- If a commit string contains new-lines, commit it in several chunks and forward Return between the chunks
- If a candidate contains newlines, replace the newlines with arrows in the lookup table
- Elide extremely long candidates in the lookup table in the middle

So something like “Best regards,\n\n\nPaul” is displayed as “Best regards,↩︎↩︎↩︎Paul” in the lookup table.

And if an entry in the lookup table is longer than 80 characters only the first 40 and the last 40 characters are shown with eliding dots in the middle.

**Comment by psads-git** [comment]
That is fine, Mike! Thanks! Will a rpm of ibus-typing-booster be meanwhile created and made available to dnf? If not, could you please direct me to some documentation where I could find information about how to install ibus-typing-booster from source?

**Comment by mike-fabian** [comment]
I think you are using Fedora 33 so I build a test rpm for you here:

https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/build/2063136/

https://download.copr.fedorainfracloud.org/results/mfabian/ibus-typing-booster/fedora-33-x86_64/02063136-ibus-typing-booster/

This is not yet the "real" 2.10.6 release, just a release candidate, I might add other stuff before the release.

The setup tool is not yet modified to allow multiline shortcuts, so you still have to paste from gedit.

**Comment by mike-fabian** [comment]
With the rpm from the last comment, it looks like this:

![Screenshot](https://user-images.githubusercontent.com/2330175/110624673-d39c8c00-819e-11eb-993a-b2667c5fb7ea.png)

**Comment by mike-fabian** [comment]
And I think it should work for you now in GMail in firefox, at least it worked for me with this update.
And it will probably work in Thunderbird, I didn't test that, but maybe you can test that.

**Comment by mike-fabian** [comment]
You can notice the small 100ms delay after each line though. If you have a shorcut which is 10 lines long, inserting that shortcut takes about a second. Unfortunately that hack with the delay is necessary to make it work. And there may be situations, maybe when the system is under very high load, that 100ms are not enough. I am not sure, for me this works reliably. I use a similar delay in ibus-typing-booster for a long time already in other places. For example, if you type a word, then use the arrow left key to go to the middle of the word (while the word is still in preedit and not committed), then type "Return". In that case, the part of the word to the left of the cursor is committed, then a Return key is forwarded, then the part of the word to the right of the cursor is forwarded. In between I also needed to insert delays to make that work reliably. But this never failed for me so far, so I think it works and the only disadvantage is that inserting shortcuts which huge expansions takes noticable time.

**Comment by psads-git** [comment]
Thanks, Mike. The delay is not noticeable at all. Great job!

**Comment by mike-fabian** [comment]
I just improved the setup tool to make it possible to write multi-line expansions there directly, without pasting from gedit.

New rpm coming in a few minutes....

**Comment by mike-fabian** [comment]
New build with the improved setup tool:

https://download.copr.fedorainfracloud.org/results/mfabian/ibus-typing-booster/fedora-33-x86_64/02063989-ibus-typing-booster/

Looks like this:

![Screenshot](https://user-images.githubusercontent.com/2330175/110659492-ec1e9d80-81c2-11eb-8777-b090badc7a36.png)

**Comment by mike-fabian** [comment]
Note that when you press the "Add shortcut" button, I remove whitespace from the beginning *and* the end of the text found in the buffer which contains the expansion.

I am not sure about it. I thought that especially the white space at the end might be confusing, as you cannot see it.

But removing the leading and trailing white space makes it impossible to have an expansion which adds a newline or spaces at the end.

Should I allow leading and trailing white space?

**Comment by mike-fabian** [comment]
> 1. To make size-adjustable the window where the user writes the expansion text.

Good idea, I may do that later as it is a bit more effort.

I did  the following small improvement which was easy now:

- make the window for the expansion text vertically expandable
- make the button area at the bottom *not* expandable.

That way, the window for the expansion text and the window for the already defined shortcuts are the only vertically expandable things in that page, if you resize the setup  tool, both will get bigger. The area for the "About" and "Close" buttons at the bottom used to get bigger when resizing the setup tool but not anymore in order not to waste useful space on these buttons.

**Comment by mike-fabian** [comment]
> 2. To allow the edition of saved shortcuts.

You know that you can define more than one different expansion for the same shortcut?

Your suggestion is useful though, I am thinking about how to do that best.

**Comment by mike-fabian** [comment]
Somewhat bigger expansion window and somewhat resizable by resizing the whole setup tool:

https://download.copr.fedorainfracloud.org/results/mfabian/ibus-typing-booster/fedora-33-x86_64/02067353-ibus-typing-booster/

**Comment by psads-git** [comment]
Now, Mike, it is perfect! Thanks for your efforts to make our life so much easier through ibus-typing-booster and its improvement!

**Comment by mike-fabian** [comment]
I will try to make the resizing work better eventually by making it possible to move a border between the upper and lower window up and down with the mouse. But that is difficult and I don’t know how to do it at the moment so I will not do this in the next month. But later probably, if I figure out how.

**Comment by psads-git** [comment]
That is very fine, Mike: it is an improvement of not high priority.

**Comment by mike-fabian** [comment]
> To allow the edition of saved shortcuts.

Can you open a new issue for this one, so that it does not get forgotten?

**Comment by mike-fabian** [comment]
Fedora 33 build of ibus-typing-booster 2.11.0 is here:

https://download.copr.fedorainfracloud.org/results/mfabian/ibus-typing-booster/fedora-33-x86_64/02068370-ibus-typing-booster/

**Comment by psads-git** [comment]
Thanks, Mike!

**Comment by mike-fabian** [comment]
I improved the behaviour of the buttons “Clear input”, “Delete shortcut”, and “Add shortcut” buttons a bit.

Now the delete and add buttons are only sensitive when both the shortcut entry and the expansion entry contain some none-whitespace text. And the clear button is only sensitive if at least  one of them is not empty (whitespace only counts as not empty).

Before the buttons were always sensitive.

**Comment by mike-fabian** [comment]
The list of already defined shortcuts at the bottom is currently ordered  in the order they were defined.

I thought alphabetical might be better, not sure  ...

**Comment by psads-git** [comment]
Thanks, Mike! If it does not imply too much work, it would be best if the shortcuts were (user) sortable by each field.

**Comment by psads-git** [comment]
There is something that it strange (maybe a bug): shortcuts that begin with numbers never seem to be suggested. For instance,

1a

to expand to

One artist.

**Comment by mike-fabian** [comment]
> Thanks, Mike! If it does not imply too much work, it would be best if the shortcuts were (user) sortable by each field.

Can you open a new issue for that to remind me?

**Comment by mike-fabian** [comment]
> There is something that it strange (maybe a bug): shortcuts that begin with numbers never seem to be suggested. For instance,
> 
> 1a
> 
> to expand to
> 
> One artist.

That is not really a bug.  

The reason for this is that with the default key bindingd, “1” and “KP_1” (The 1 one the number pad if your keyboard has a number pad) commit the first candidate and a space. These are the default key bindings:

![default-keybindings](https://user-images.githubusercontent.com/2330175/111033668-a3f5ba00-8412-11eb-8608-3df5de90fd27.png)

**Comment by mike-fabian** [comment]
Now if “1” is used to commit a candidate you cannot type it into the preedit.  For example, if you type “a” you have “a” in the preedit and see some completions starting with “a” in the candidate list. If you then type a “1” the first of these candidates is committed, you do not get “a1” in the preedit.


Let’s assume the preedit is empty, and you type a “1”. Why not put this into the preedit? There are no candidates yet, so there is no conflict with committing a candidate? So putting it into the preedit should be OK, no? No, it is not OK because as soon as the 1 is in the predit are most likely candidates and now you cannot type the next digit. So if you want to type the number eleven “11” and the first “1” went into the preedit, the second “1” would commit something and you would not get the number eleven you wanted to type.

That means, if digits like “1” are used for any keybindings (especially the key bindings to commit candidates but also some othes would be problematic), then they should not start a preedit to make it still possible to type numbers.

So if the preedit is empty and the first key is typed, and it is a number, and numbers are used in any keybindings, then ibus-typing-booster doesn’t open a preedit and puts that number into the preedit but rather commits that number directly (possibly converted to a localised number if you use Arabic or Indic input methods ...).

But there is one possibility to change this behaviour, maybe it is enough for you to use F1…F9 to commit candidates?  You don’t need numbers in any keybindings at all maybe?

Then you can remove all numbers from the keybindings and then you can input them into the preedit and you can type “1a” into the preedit and it can complete or expand as a custom shortcut.

That means you could make the keybindings like this, and they you can type the numbers  into the preedit:

![keybindings](https://user-images.githubusercontent.com/2330175/111034315-56c71780-8415-11eb-9f26-bb01943b96f9.png)

**Comment by mike-fabian** [comment]
The code which does this is here:

https://github.com/mike-fabian/ibus-typing-booster/blob/master/engine/hunspell_table.py#L1417

```
        # Update self._digits_used_in_keybindings:
        self._digits_used_in_keybindings = False
        for command in keybindings:
            for keybinding in keybindings[command]:
                if (keybinding
                    in ('0', '1', '2', '3', '4',
                        '5', '6', '7', '8', '9',
                        'KP_0', 'KP_1', 'KP_2', 'KP_3', 'KP_4',
                        'KP_5', 'KP_6', 'KP_7', 'KP_8', 'KP_9')):
                    self._digits_used_in_keybindings = True
```

and here:

https://github.com/mike-fabian/ibus-typing-booster/blob/master/engine/hunspell_table.py#L4968

```
            if self.is_empty():  # <- that means the preedit is empty!

                [...]
                    
                if (self._digits_used_in_keybindings
                    and not self._tab_enable
                    and key.msymbol
                    in ('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')):
                    # If digits are used as keys to select candidates
                    # it is not possibly to type them while the preëdit
                    # is non-empty and candidates are displayed.
                    # In that case we have to make it possible to
                    # type digits here where the preëdit is still empty.
                    # If digits are not used to select candidates, they
                    # can be treated just like any other input keys.
                    #
                    # When self._tab_enable is on, the candidate list
                    # is only shown when explicitely requested by Tab.
                    # Therefore, in that case digits can be typed
                    # normally as well until the candidate list is
                    # opened.  Putting a digit into the candidate list
                    # is better in that case, one may be able to get a
                    # reasonable completion that way.

                [... commit the digit immediately or transliterate it and commit it ...]
```

**Comment by mike-fabian** [comment]
As you  can see in the comment in the second chunk of code, there is another way you could type digits into the preedit without removing them from all keybindings:

You could use the option “Enable suggestions by key (Default is the Tab key)”.

If that option is use, you only get suggestions on request. This is useful for user who don’t want to complete stuff all the time and find the candidate list popping up all the time annoying. But from time to time they want a candidate list because they are nor sure about the spelling of a word, want to input an emoji, want to expand a custom shortcut ...

With that option on, “1a” goes into the preedit and no candidate list pops up. Typing Tab then  pops up the candidate list which contains “one artist” if you defined “1a” as a shortcut for this.

**Comment by psads-git** [comment]
> > Thanks, Mike! If it does not imply too much work, it would be best if the shortcuts were (user) sortable by each field.
> 
> Can you open a new issue for that to remind me?

Done so meanwhile, Mike. Thanks!

**Comment by mike-fabian** [comment]
If you look a the code chunks above,  you see that the decision whether number are allowed in the preedit or not is a bit rough.

When only a single digit, whether from the digit row or the keypad digits is in any keybinding, digits are not allowed into the preedit and are committed (possibly transliterated) directly.

This could be improved to more fine grained control.

For example, if the keyboard used has both a number row *and* a numeric keypad, then one could decide to commit candidates only with the digits from the number row but not commit candidates from the numeric keypad. 

I.e. change the keybindings to something like:

```
commit_candidate_1_plus space   ['1', 'F1']
```

Then, if I implemented more fine grained control, I could allow digits typed using the numeric keypad to enter the preedit now because they are not used in keybindings anymore.

Or the other way around, if the keybindings were changed to

```
commit_candidate_1_plus space   ['KP_1', 'F1']
```

then I could allow digits typed with the digit row into the  preedit and candidates would only be committed when typing the digits with the numeric keypad. 

I would probably implement this based on individual keys if I do it.

I.e. if the preedit is empty and a digit is typed using  0, 1, ..., 9, KP_0, KP_1, ... , KP_9, then check whether this particular key to type a digit is used in any keybinding and if not allow it into the preedit and if yet commit it.

**Comment by mike-fabian** [comment]
If you think I should implement that, please open an issue for that.

**Comment by mike-fabian** [comment]
By the way, probably from Monday on I will be away until the beginning of April so don’t be surprised if I cannot reply then.

**Comment by psads-git** [comment]
> If you think I should implement that, please open an issue for that.

I shall open an issue for that, yes, Mike. Thanks!

**Comment by psads-git** [comment]
> By the way, probably from Monday on I will be away until the beginning of April so don’t be surprised if I cannot reply then.

No problem, Mike! Enjoy!

**Comment by psads-git** [comment]
The problem is not only with shortcuts. The following just happened to me:

Mac M1

expands to

Mac Mac.

I would expect the same problem while writing down telephone numbers.

**Comment by mike-fabian** [comment]
> The problem is not only with shortcuts. The following just happened to me:
> 
> Mac M1
> 
> expands to
> 
> Mac Mac.
> 
> I would expect the same problem while writing down telephone numbers.

Yes, of course, if 1 commits the first candidate, you cannot write M1 into the preedit.

If you want to be able to do that, remove the keys 1, ..., 9, KP_1, ..., KP_9 from the keybindings. 

But then you have to type F1 to commit. And on most keyboards the F1 key is further away from the fingers than the 1 key, so that might not be convenient.

Of course one can commit using the mouse but that is even less convenient because one hand has to leave the keyboard to grab the mouse. 

Actually the way I personally most often select and commit a candidate is by hammering on the Tab key and then space.
First hit of the Tab key selects the first candidate, second Tab key the second candidate, ... when the right candidate is reached commit with space (Shift+Tab goes backwards by the way).

So by default the digits 1, ..., 9, KP_1, ..., KP_9 are used to select and commit candidates, but there are other options and if one is happy with one of the other options one can remove the digits from the keybindings and then become able to type digits into the preedit.

If I improve the current code I can make it possible leave *either* 1, ..., 9 *or* KP_1, ..., KP9 in the keybindings to commit candidates and use the other which is not in the keybindings to type digits into the preedit. Currently it is all or nothing, currently you have to remove *all* digit keys from the keybindings to be able to type digits into the preedit. I can improve it so make it possible to type digits into the preedit with one set of digit keys and commit candidates with the other set of digit keys.

**Comment by mike-fabian** [comment]
https://download.copr.fedorainfracloud.org/results/mfabian/ibus-typing-booster/fedora-33-x86_64/02074075-ibus-typing-booster/

This build contains the above change plus one other change, this one:

commit fdf1b8b84c6c9ad34dceaf91a6366b3f28348950
Author: Mike FABIAN <mfabian@redhat.com>
Date:   Fri Mar 12 16:53:14 2021 +0100

    Don’t use surrounding text if InputPurpose.TERMINAL is set.
    
    I don’t know of any terminal which has working surrounding text support.
    xfce-4-terminal and gnome-terminal claim to have surrounding text support
    by setting the respective client capabilities, so ibus-typing-booster
    thinks surrounding text should work. But it doesn’t work at all in these
    terminals, it is even worse than not working at all because the surrounding
    text retrieved is from the application used before changing the focus to the terminal.
    See this bug:
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1847347
    (“gnome-terminal and xfce4-terminal get surrounding text from previous application”)
    
    But gnome-terminal and xfce4-terminal also set InputPurpose.TERMINAL.
    So I can ignore surrounding text support when InputPurpose.TERMINAL is set.
    That improves the current situation a bit for gnome-terminal and xfce4-terminal.
    Better no surrounding text at all then a completely wrong one.
    
    As soon as a terminal which sets InputPurpose.TERMINAL *and* has really
    working surrounding text support, I have to change this again, but
    for at the moment it helps.

**Comment by mike-fabian** [comment]
As you are interested in the user defined shortcut feature of ibus-typing-booster, what do you think of this?:

https://github.com/mike-fabian/ibus-typing-booster/issues/271

Is my implementation to allow punctuation in shortcuts and make shortcuts case and accent sensitive OK for you?

**Comment by psads-git** [comment]
Thanks, Mike, for asking me for my opinion. I hope you are well and in good health.

Your implementation of allowing punctuation in shortcuts is a real improvement! It allows for better discrimination among shortcuts, specially for users that define many shortcuts. Thanks!

Now, changing the subject a bit. Recently, I came across the following:

https://www.lightkey.io/

Maybe it will help us to get new ideas for ibus-typing-booster!

**Comment by mike-fabian** [comment]
> Your implementation of allowing punctuation in shortcuts is a real improvement! It allows for better discrimination among shortcuts, specially for users that define many shortcuts. Thanks!

Thank you! And what about allowing upper case and accents to distingish user shortcuts? I think this is probably OK as well but I am not 100% sure.

**Comment by mike-fabian** [comment]
> Now, changing the subject a bit. Recently, I came across the following:
> 
> https://www.lightkey.io/
> 
> Maybe it will help us to get new ideas for ibus-typing-booster!

Thank you! I’ll have a look!

**Comment by psads-git** [comment]
> Thank you! And what about allowing upper case and accents to distingish user shortcuts? I think this is probably OK as well but I am not 100% sure.

It may be useful, Mike, specially if the expansion of the shortcuts is long. For short expansions, I do not see great usefulness, since the shortcut is more costly to type (e.g., accented characters take pressing two keys in sequence). Personally, I always avoid typing accents.

And this discussion is now triggering an idea on my mind: Should not misspelled words be automatically corrected? For instance, when one writes

`tambem`

(word that does not exist in Portuguese), should not ibus-typing-booster correct that to

`também`

?

**Comment by mike-fabian** [comment]
> > Thank you! And what about allowing upper case and accents to distingish user shortcuts? I think this is probably OK as well but I am not 100% sure.
> 
> It may be useful, Mike, specially if the expansion of the shortcuts is long. For short expansions, I do not see great usefulness, since the shortcut is more costly to type (e.g., accented characters take pressing two keys in sequence). Personally, I always avoid typing accents.

Yes, I probably would not use accents in my user shortcuts either. Maybe I would use capital letters, but probably not accents.

For normal words we investigated that saving the capital letters and/or the accents to the user database was not good and made the prediction quality on average worse.

For normal words, we also strip the punctuation before saving them to the user database, saving them together with the punctuation would also make the prediction quality worse.

Before I implemented the improvement for the user shortcuts, the user shortcuts were saved to the user database with the same restrictions as any other normal word, i.e. punctuation was strippe, everything was converted to lower case and accents were stripped.

But while changing the code to allow punctuation in user shortcuts, I noticed that I could just as well also remove the limitation not to use capital letters and accents because user shortcuts are a special case anyway. Although saving to the user database without punctuation, capital letters and accents improves the prediction for normal words, this statistical improvment doesn't seem to apply at all to user shortcuts.

So I thought there is no harm in making all of these (punctuation,  capital letters, and accents) possible in user shortcuts. Might be useful to somebody and won't cause any harm.

I just had some remaining doubts and if I release it like that and later discover that allowing capital letters and accents in user shortcuts was not a good idea and I change this again, then the users who did define such shortcuts with capital letters and/or punctuation would lose these shortcut again if the update removes the support. Quite unlikely, I don't think it causes any harm so I probably will never remove that.

**Comment by psads-git** [comment]
Yes, Mike, I agree that it is quite unlikely that it will be needed to remove accented and capitalized user shortcuts as an ibus-typing-booster feature. So, I guess that you can go ahead.

**Comment by mike-fabian** [comment]
> And this discussion is now triggering an idea on my mind: Should not misspelled words be automatically corrected? For instance, when one writes
> 
> `tambem`
> 
> (word that does not exist in Portuguese), should not ibus-typing-booster correct that to
> 
> `também`
> 
> ?

If you type `tambem` while using a Portuguese dictionary, you will already get `também` as a candidate. It is even the first candidate by default unless your previous typing has made Typing Booster learn something different.

So you can just select the first candidate (e.g. by typing Tab) and commit (e.g. with space). Or select and commit in one go by typing `1` or `F1`.

So ibus-typing-booster already does correct this, just not 100% automatically.

ibus-typing-booster also already has an option

```
[✓] Automatically select the best candidate
```

If that option is on and you type `tambem` and `também` is the first candidate, it is already selected (blue background in the candidate list) and space commits it.

I.e. typing `tambem` + space would commit `também `.

SwiftKey on Android has a similar option

```
[✓] Autocorrect (Words will be corrected to the middle prediction as you type)
```

(Middle prediction is the middle one of the three prediction buttons on the smartphone, the middle prediction is the “best” prediction, i.e. this is what is the first candidate in ibus-typing-booster.)

I hate this option and always switch it off. If this is on, I always get some some automatic “spelling corrections” inserted while I just continue to type and don’t select any prediction candidates manually. And they are almost always wrong for me, they introduce lots of silly and sometimes funny mistakes but rarely help.

Sometimes I think adding the

```
[✓] Automatically select the best candidate
```

to Typing Booster was a mistake, for me it causes more harm than good.

But you can try to experiment with that option and see whether it does anything useful for you.

I.e. the feature you ask for already exists, misspelled words can be automatically corrected already by using that option, if there is any further work to be done it would be to improve the details of this automatic correction.

Spelling corrections are in most cases not unique, in most cases the spellchecking programs suggest many possibilities, for example:

```
$ echo carmel | hunspell -d en_US 
Hunspell 1.7.0
& carmel 6 0: calmer, camel, caramel, carrel, cartel, carpel
```

I you wanted to write `camel` but accidentally typed `carmel` and then get `calmer` committed just because it is the first candidate suggested by the spellchecker isn't really helpful.

At the moment, I find the option

```
[✓] Automatically select the best candidate
```

pretty useless. Maybe it could be useful if it would select the best candidate **not** **always** but only if this best candidate is really **extremely** **good**.

And we would need to figure out what **extremely** **good** means.

Maybe it could mean something like: “It has all the same base letters in the same order and only differs by accents”. That would select `também` automatically when typing `tambem`.

I could just change the behaviour of that option a to not blindly select the first candidate always but only in really good cases.

Or, I could change the checkbox option `[✓] Automatically select the best candidate` into a combobox option like

```
Automatically select the best candidate [Never|For very good candidates|Always]
```

**Comment by mike-fabian** [comment]
Typing `tambem` using a Portuguese dictionary with and without the option 

```
[✓] Automatically select the best candidate
```
https://user-images.githubusercontent.com/2330175/155746663-a14c4dcd-c640-43f2-8b7a-8fdfb830dc15.mp4

**Comment by psads-git** [comment]
Thanks, Mike! I have been aware of the option

`[✓] Automatically select the best candidate`

And, like you, I do not use such an option, because it is not enough intelligent. 

What I was meaning was something along what you discuss in the last part of your message. Sometimes, it can be determined, with almost full certainty, what the candidate selection should be: The example of word `tambem`. In factt, that is a word that does not exist in the Portuguese language and, moreover, almost coincides with a word belonging to the Portuguese dictionary: `também`. I believe that, at least in Portuguese, there are numerous words in the class I have just defined.

**Comment by mike-fabian** [comment]
> Thanks, Mike! I have been aware of the option
> 
> `[✓] Automatically select the best candidate`
> 
> And, like you, I do not use such an option, because it is not enough intelligent.
> 
> What I was meaning was something along what you discuss in the last part of your message. Sometimes, it can be determined, with almost full certainty, what the candidate selection should be: The example of word `tambem`. In fact, that is a word that does not exist in the Portuguese language and, moreover, almost coincides with a word belonging to the Portuguese dictionary: `também`. I believe that, at least in Portuguese, there are numerous words in the class I have just defined.

Yes, that is the exactly the case I meant with third possible option in 

```
Automatically select the best candidate [Never|For very good candidates|Always]
```

`Never` would be the same as if the currently available option is off.

`Always` would be the same as if the currently available option is switched on.

`For very good candidates` would automatically select only in such cases where we have “almost full certainty”.

We might have “almost full certainty” if the typed word and the first candidate have the same length, all base letters are the same and the order of the letters is also the same. The only difference is that the typed word has no accents but the first candidate has accents. I guess in that case we have “almost full certainty” that this is correct (There maybe some corner cases where it is still incorrect, but probably not many).

**Comment by mike-fabian** [comment]
> > Thanks, Mike! I have been aware of the option
> > `[✓] Automatically select the best candidate`
> > And, like you, I do not use such an option, because it is not enough intelligent.
> > What I was meaning was something along what you discuss in the last part of your message. Sometimes, it can be determined, with almost full certainty, what the candidate selection should be: The example of word `tambem`. In fact, that is a word that does not exist in the Portuguese language and, moreover, almost coincides with a word belonging to the Portuguese dictionary: `também`. I believe that, at least in Portuguese, there are numerous words in the class I have just defined.
> 
> Yes, that is the exactly the case I meant with the third possible option in
> 
> ```
> Automatically select the best candidate [Never|For very good candidates|Always]
> ```
> 
> `Never` would be the same as if the currently available option is off.
> 
> `Always` would be the same as if the currently available option is switched on.
> 
> `For very good candidates` would automatically select only in such cases where we have “almost full certainty”.
> 
> We might have “almost full certainty” if the typed word and the first candidate have the same length, have a minimum length of say 5 letters, all base letters are the same and the order of the letters is also the same. The only difference is that the typed word has no accents but the first candidate has accents. I guess in that case we have “almost full certainty” that this is correct (There maybe some corner cases where it is still incorrect, but probably not many).

So instead of adding a new option, I would just add an additional possible setting for the already existing option.

**Comment by psads-git** [comment]
Yes, Mike, it would not be a new option, but a level of the already existent option.

That may be implemented without any programming: By using shortcuts! For instance, a shortcut for `também` could automatically be expanded to `também`. I am not sure what is the best approach.

**Comment by psads-git** [comment]
Actually, Mike, I think that maybe an option for automatic expansion of user-selected shortcuts should be added, as some shortcuts are candidates with full certainty.

**Comment by psads-git** [comment]
Maybe, Mike, we can go even further with this idea of "almost full certainty": Perhaps, if the first candidate has a ranking too above the remaining candidates, the first candidate is an "almost full certainty" candidate!

**Comment by mike-fabian** [comment]
> Maybe, Mike, we can go even further with this idea of "almost full certainty": Perhaps, if the first candidate has a ranking too above the remaining candidates, the first candidate is an "almost full certainty" candidate!

Yes.

**Comment by mike-fabian** [comment]
It sounds like a pretty good idea.

Not very easy to implement, I think I might have time for this in April.

**Comment by psads-git** [comment]
Thanks, Mike! There is no hurry: Implement that only when you can.

I think it will also lead to a cleaner database, as, in my case, with word `tambem`, for instance, I prefer to commit it and correct that after I finish my phrase.

Maybe, to define the ranking of the candidates, a classification machine learning model can be used.

**Comment by mike-fabian** [comment]
> I think it will also lead to a cleaner database, as, in my case, with word `tambem`, for instance, I prefer to commit it and correct that after I finish my phrase.

As we save “normal” words to the database without accents, it does not matter here. “também” is always saved as “tambem” to the database.

**Comment by psads-git** [comment]
You are right, Mike: My example was badly chosen! But sometimes one writes with typos and prefers to wait to correct the type after having finished the phrase. Suppose that one writes `tamben` (notice the `n` instead of the `m`): With automatic correction, the database would get cleaner.

**Comment by mike-fabian** [comment]
> You are right, Mike: My example was badly chosen! But sometimes one writes with typos and prefers to wait to correct the type after having finished the phrase. Suppose that one writes `tamben` (notice the `n` instead of the `m`): With automatic correction, the database would get cleaner.

Yes, but in a case like `tamben` versus `também` it is not clear anymore whether this is really “almost full certainty”.

For example in German, `grüner`, `grünes`, `grünem`, `grünen` are all valid words.

**Comment by psads-git** [comment]
I guess, Mike, that in the dictionary there is no word closer to `tamben` than `também`:

http://todaspalavras.com/starts-with-by-length/tamb/

So, this is also a case of "almost full certainty", I guess.

**Comment by mike-fabian** [comment]
> I guess, Mike, that in the dictionary there is no word closer to `tamben` than `também`:
> 
> http://todaspalavras.com/starts-with-by-length/tamb/
> 
> So, this is also a case of "almost full certainty", I guess.

Yes, but it is more complicated than that.

One complication is that one could be using several languages the same time for example. Then you have to consider whether there is no closer word in several languages at the same time.

Hunspell isn’t so sure that there isn’t anything closer to `tamben` than `também` either.
When asked for spellchecking suggestions for `tamben` in Portuguese, Hunspell gives:

```
[mfabian@fedora ~]$ echo -n 'tamben' | hunspell -d pt_BR 
Hunspell 1.7.0
& tamben 12 0: tambeense, entambe, tambetá, butambeno, tambetara, sambenita, intambe, mutambense, tambeiro, tambetaru, tatambear, sambento

[mfabian@fedora ~]$
```

`também` isn’t even in the list of suggestions.

Even though Hunspell knows that `também` is a valid Portuguese word:

```
[mfabian@fedora ~]$ echo -n 'também' | hunspell -d pt_BR 
Hunspell 1.7.0
*

[mfabian@fedora ~]$
```

No suggestions here means `também` is a correctly spelled word.

**Comment by psads-git** [comment]
I thought Hunspell more accurate than in reality is. For instance, Microsoft Word guesses that `tamben` may correspond to `também`:

![image](https://user-images.githubusercontent.com/75945439/155792232-4dfea25b-f97b-4391-9edc-b9236790069c.png)

**Comment by mike-fabian** [comment]
> One complication is that one could be using several languages the same time for example. Then you have to consider whether there is no closer word in several languages at the same time.

![Screenshot](https://user-images.githubusercontent.com/2330175/155950971-263d662e-4685-435e-a52a-5f2744fddb8a.png)

**Comment by mike-fabian** [comment]
In the example in my last comment I am learning French with duolingo. I have to type both French and English sentences. Therefore, I have setup Typing Booster to use both the English and the French dictionary.

In the above screenshot, I want to type “arrived”. As the French dictionary is also active, I get “arrivé” as the first candidate (Actually it is not even from the French dictionary but already from previous user input, I can see that because of the black colour and the missing dictionary symbol 📖). If I had typed the same context “would have arrived” recently, the first candidate would probably be “arrived” already, but apparently I didn’t type that yet. 

So I get the French suggestion “arrivé”. This first candidate has the same base letters in the same order as the typed word and only adds accents. If we conclude “almost full certainty” now, then “arrivé” would be automatically selected. If I notice that,I could deselect it by typing `Escape`. But if I don’t notice and just continue typing the final `d` of `arrived`, then the selected “arrivé” gets committed and the `d` added. So the final result becomes “arrivéd”. And then I have a weird error in my final sentence:

“Without this traffic jam we would have arrivéd already”

This is the reason why I always switch the option

```
[✓] Autocorrect (Words will be corrected to the middle prediction as you type)
```

off when using SwiftKey on Android because it very often introduces such errors for me.

Similar things could happen with tambem/também maybe.

So it is really hard to figure out whether the first candidate  really has “almost full certainty” to be correct or not.

**Comment by psads-git** [comment]
I am not still fully convinced, Mike. Take your example: 

`Without this traffic jam we would have arrivéd already`

The trigram: 

`would have arrivéd`

would have zero frequency of previous use. Then, in your example, of all 6 suggestions, only the correct one would qualify as “almost full certainty” candidate, as all the others would have a trigram with zero frequency. 

Moreover, most of people use only a single dictionary or maybe two at most.

**Comment by mike-fabian** [comment]
I didn’t want to say that it is impossible to figure out when one really has “almost full certainty”, just that it is very difficult.

My example only showed that the simplistic rule “spelling is the same only accents  are missing” is not good enough.

As you say one also needs to consider the frequency of previous use of the whole context.

**And** one probably should consider whether words are found in dictionaries for the whole context, not only for the current word.

For example, when one has typed “would have arrive”, then the last word “arrive” is found in both the French **and** the English dictionary but “would” and “have” are only found in the English dictionary. So the “arrivé” is definitely not a “full certainty” suggestion, even though it just adds an accent. As “arrive” without the accent is also a valid French word, even  if the  whole  context is French it isn’t necessarily correct to add the accent. 

“Le train arrive” = “The train arrives”
“Le train à arrivè” = “The train has arrived” 

So for “almost complete certainty” we probaly need something like:

1. the whole context can be found in a dictionary of a single language
2. the whole context has a high user score because it has been typed many times before
3. the suggestion doesn’t change the order  of any letters
4. the suggestion doesn’t change the length, only adds or removes accents.

I guess 1. and 2. are really a necessary condition for a “almost complete certainty” suggestion. I am not sure about 3. and 4.

But overall it is a quite difficult problem, I have to think about it and try out several things.

Such a feature has to work correctly almost 100% of the time, if it introduces too many errors it is not helpful (like the current “Automatically select the best candiate” option is almost  useless).

On the other hand, if the conditions to have “almost complete certainy” are so strict that they are almost never met, then such a feature would not help much either. It would help all that much if you get an automatic selection only once in a blue moon when the stars are aligned perfectly. 

I think I will probably try to add third value to that option to autoselect the best candidate **if** we have “almost complete certainty” for that candidate. But I think it is not easy at all.

**Comment by psads-git** [comment]
I think the 4 conditions you, Mike, enumerate for "almost full certainty" are good -- they will very rarely fail to predict the word wanted by the user.

The automatic selection will not _for sure_ be of the type "only once in a blue moon". Why? Because that will allow the user to write completely without accents. Thus, automatic selection will always be in use, increasing typing speed a lot while providing a pleasant experience for the user.

**Comment by mike-fabian** [comment]
I will certainly try that when I come back from vacation.

**Comment by psads-git** [comment]
Thanks, Mike. Have a nice vacation!

**Comment by mike-fabian** [comment]
I opened a new issue for this: https://github.com/mike-fabian/ibus-typing-booster/issues/285
